lib/ops/opsservice/agents.go

```diff
@@ -394,6 +396,10 @@ func (r *AgentPeerStore) authenticatePeer(token string) (*storage.ProvisioningTo
 
 func (r *AgentPeerStore) validatePeer(ctx context.Context, group *agentGroup, info storage.System,
 	req pb.PeerJoinRequest, clusterName string) error {
+	if group.hasPeer(req.Addr, info.GetHostname()) {
```

Comment:
Hmm, I'm not sure about this. Let's say I have two VMs with the same hostname (say, `node-1`) and the same internal IP, which is not uncommon (e.g. all my local VMs have IP `10.0.2.15`). With this change they will both be allowed to join the installer?

Reply:
In this case we need another differentiator apart from the tuple (address, hostname) as this is not going to work regardless of this change. What this is fixing is the race between the usual agent reconnect and the test connection that the agent sends at the very beginning (which is the difference between 5.5.x and 5.0.x) to verify the possibility of the reconnect.
Besides, an agent can drop off any time (without triggering `PeerLeave` - i.e. due to temporary networking issue) and then reconnect (effectively resending the `PeerJoin`) so it is supposed to be idempotent.

Reply:
Another question regarding your VM setup - how do set up routing between the VMs or VM and the host on that IP if it's the same for all of them?

Reply:
My VMs also have another host-only adapter for VM-to-VM communication but that was just an example - I was mostly wondering how we could detect similar situation and prevent a user from making such mistake? This is definitely an edge case though.

Just as an idea, if (ip, hostname) is not enough, could we add a mac address too? I think it can be spoofed as well but it has to be intentional and would protect against mistakes. This might be out of scope of this particular issue though.